posthog/temporal/data_modeling/run_workflow.py

```diff
@@ -415,11 +423,16 @@ async def materialize_model(
             raise CannotCoerceColumnException(f"Type coercion error in model {model_label}: {error_message}") from e
         else:
             saved_query.latest_error = f"Failed to materialize model {model_label}"
+            error_message = "Your query failed to materialize. If this query ran for a long time, try optimizing it."
             await database_sync_to_async(saved_query.save)()
+            await mark_job_as_failed(job, error_message)
             raise Exception(f"Failed to materialize model {model_label}: {error_message}") from e
 
-    tables = get_delta_tables(pipeline)
+    data_modeling_job = await database_sync_to_async(DataModelingJob.objects.get)(id=job.id)
+    if data_modeling_job.status == DataModelingJob.Status.CANCELLED:
+        raise DataModelingCancelledException("Data modeling run was cancelled")
```

Comment:
If we cancelled the workflow while running `await asyncio.to_thread(pipeline.run, ...)`, then a `CancelledError` exception will be thrown and the `DataModelingJob` would be marked as `FAILED` by the exception handling code. This can lead to a race condition with the `cancel` method of the saved query view. Depending on who writes to the db first, you may end up with a `FAILED` here.

Reply:
Also, if we finished already, I don't think we should fail here necessarily. If it's the last model, and we are done, who cares we are cancelled? We succeeded already.

Reply:
Ok, that seems reasonable enough, one of the tricky items for me here is that when you cancel a job, then it shows up as completed because it actually already finished by the time the cancellation request was received. Anyway, I will play around with this other direction.

Reply:
Hmm, yeah, I feel like if the job completed we shouldn't error, even if a cancellation request was issued. We already did the work, not really a point in throwing it all away.